src/main/java/org/truffleruby/core/array/ArrayNodes.java

```diff
@@ -551,22 +551,25 @@ public DynamicObject clear(DynamicObject array,
     @ImportStatic(ArrayGuards.class)
     public abstract static class CompactNode extends ArrayCoreMethodNode {
 
-        @Specialization(guards = { "strategy.matches(array)", "!strategy.accepts(nil())" }, limit = "ARRAY_STRATEGIES")
+        @Specialization(guards = { "strategy.matches(array)", "strategy.isPrimitive()" }, limit = "STORAGE_STRATEGIES")
         public DynamicObject compactPrimitive(DynamicObject array,
                 @Cached("of(array)") ArrayStrategy strategy) {
             final int size = strategy.getSize(array);
             Object store = strategy.newMirror(array).extractRange(0, size).getArray();
+            strategy.setStore(array, strategy.newMirror(array).extractRange(0, size).getArray());
```

Comment:
This should use "unshareArray()" or so, so it's a no-op if strategy is not a copy-on-write strategy.
But I'm a bit confused, what do we gain with sharing storage here?
Wouldn't it be simpler to take a full copy of the storage and not change the source `array`?

Reply:
@aardvark179 Could you clarify?